### ISSUE TITLE
topic.partitionMetadata.forEach is not a function

### ISSUE THREAD

**Issue by scmorse** [oleksiyk/kafka#121]
Within `no-kafka/lib/client.js`, `Client.prototype.updateMetadata` includes these lines:

```
_.each(response.topicMetadata, function (topic) {
    self.topicMetadata[topic.topicName] = {};
    topic.partitionMetadata.forEach(function (partition) {
        self.topicMetadata[topic.topicName][partition.partitionId] = partition;
    });
});
```

For me, though,`topic.partitionMetadata` is an object with keys '0', '1', ..., instead of an actual array, so it does not inherit the 'forEach' method. And so, `[TypeError: topic.partitionMetadata.forEach is not a function]` is thrown.

My kafka service is [flozano/kafka:0.9](https://hub.docker.com/r/flozano/kafka/tags/). 

Wrapping `topic.partitionMetadata` with `_.toArray()` seems to fix the problem. I've been using no-kafka for a while now, and just started having this problem recently, which is strange.

**Comment by oleksiyk** [comment]
`topic.partitionMetadata` is an array. Check your code for something that mutates it.

**Comment by fhemberger** [comment]
I ran into the same issue this week:

**Node.js Version: 6.11.0** (on Alpine Linux)
**Kafka Version: 0.10.1.1**
Dockerfile: [solsson/kafka-persistent:0.10.1.1](https://hub.docker.com/r/solsson/kafka-persistent/) ([GitHub](https://github.com/solsson/kubernetes-kafka/tree/master/docker-kafka-persistent))

**Topic-Configuration:**
Partitions: 20
Replication: 3

**Code:**
```js
const producer = new Kafka.Producer({
  connectionString: 'kafka:9092'
})

producer.init()
```

I wrapped the `forEach` method above in `no-kafka/lib/client.js` with `Array.isArray(topic.partitionMetadata)` and logged everything that didn't match:

```
typeof topic.partitionMetadata
// => object

topic.partitionMetadata
// => { '0':
   { error: null,
     partitionId: 17,
     leader: 1,
     replicas: { '0': 0, '1': 1, '2': 2 },
     isr: { '0': 1, '1': 2, '2': 0 } },
  '1':
   { error: null,
     partitionId: 8,
     leader: 1,
     replicas: { '0': 0, '1': 1, '2': 2 },
     isr: { '0': 1, '1': 2, '2': 0 } },
  '2':
   { error: null,
     partitionId: 11,
     leader: 1,
     replicas: { '0': 0, '1': 1, '2': 2 },
     isr: { '0': 1, '1': 2, '2': 0 } },
  …
  }
```

@oleksiyk Any idea what might cause this?

**Comment by oleksiyk** [comment]
@fhemberger Are you able to make a standalone test that replicates the problem?

**Comment by fhemberger** [comment]
Okay, looks like I found the reason: It's not at all related to `kafka` at all.

We are using your library in a hapi web-service, with [`hapi-and-healthy`](https://github.com/atomantic/hapi-and-healthy) for health checks. At first, it sounds **really** far off that this might be causing the issue, but as soon as I removed the plug-in, everything worked as expected again. I didn't went into too much detail, but there's some patching/extending `lodash` going on (for example using [`undermore`](https://github.com/atomantic/undermore)), which might interfere with the lodash's `_.each` method.
